File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/line/call/CallIndexerTest.java
package com.flat.simplex.parser.logic.line.call;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.block.BlockIf;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class CallIndexerTest {

    @Test
    public void loadSingleIndexer() {
        TokenChain chain = readChain("[a]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(1, call.getLines().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadDoubleIndexer() {
        TokenChain chain = readChain("[a, b]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(2, call.getLines().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadComplexIndexer() {
        TokenChain chain = readChain("[a + b]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(1, call.getLines().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadEmptyIndexer_Fail() {
        TokenChain chain = readChain("[]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(0, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.lineMissingIndexers);
    }

    @Test
    public void loadTripleIndexer_Fail() {
        TokenChain chain = readChain("[a,b,c]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(3, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.lineTooMuchIndexers);
    }

    @Test
    public void loadDoubleCommaIndexer_Fail() {
        TokenChain chain = readChain("[a,,b]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(2, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void loadEndComma_Fail() {
        TokenChain chain = readChain("[a,]");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(1, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadMissingCloser_Fail() {
        TokenChain chain = readChain("[a");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(1, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.missingCloser);
    }

    @Test
    public void loadEndCommaMissingCloser_Fail() {
        TokenChain chain = readChain("[a,");
        Context context = new Context(chain.get());
        CallIndexer call = new CallIndexer(context, chain.get());
        call.load();

        assertEquals(1, call.getLines().size(), "Invalid parameters count");

        assertErrors(context, Error.missingCloser, Error.unexpectedEndOfTokens);
    }
}

File: SimplexApp/simplex/parser/logic/line/call/CallIndexer.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const ContextSupport = require("simplex/support/ContextSupport.js");
const Error = require("simplex/parser/logic/error/Error.js");
const CallIndexer = require("simplex/parser/logic/line/call/CallIndexer.js");
const AssertError = require("simplex/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertEquals = AssertError.assertEquals;

test('loadSingleIndexer', () => {
    let chain = readChain('[a]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(1, call.getLines().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadDoubleIndexer', () => {
    let chain = readChain('[a, b]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(2, call.getLines().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadComplexIndexer', () => {
    let chain = readChain('[a + b]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(1, call.getLines().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadEmptyIndexer_Fail', () => {
    let chain = readChain('[]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(0, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.lineMissingIndexers);
});
test('loadTripleIndexer_Fail', () => {
    let chain = readChain('[a,b,c]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(3, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.lineTooMuchIndexers);
});
test('loadDoubleCommaIndexer_Fail', () => {
    let chain = readChain('[a,,b]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(2, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedToken);
});
test('loadEndComma_Fail', () => {
    let chain = readChain('[a,]');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(1, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadMissingCloser_Fail', () => {
    let chain = readChain('[a');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(1, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.missingCloser);
});
test('loadEndCommaMissingCloser_Fail', () => {
    let chain = readChain('[a,');
    let context = new Context(chain.get());
    let call = new CallIndexer(context, chain.get());
    call.load();
    assertEquals(1, call.getLines().length, 'Invalid parameters count');
    assertErrors(context, Error.missingCloser, Error.unexpectedEndOfTokens);
});